### ISSUE TITLE
Unclear how discrete executions should work with reenqueued  jobs (leads to broken job-iteration)

### ISSUE THREAD

**Issue by julik** [bensheldon/good_job#972]
We recently migrated our installation by adding the executions table, and we stumbled upon an issue with [job-iteration](https://github.com/Shopify/job-iteration)

Basically, it seems that when job-iteration re-enqueues the job with the adjusted cursor value, good_job does not consider this as a reenqueue but as an Execution which has completed. The `finished_at` gets set and execution ends. We could bypass this issue by forcing an exception to be raised when job-iteration re-enqueues the next iteration (here https://github.com/Shopify/job-iteration/blob/1bf0395fca14818fe12d846ba4aa1cf382944ee2/lib/job-iteration/iteration.rb#L222 )

What we were wondering however - and it is not very easy to figure out - is this: how is the identity _supposed_ to work with discrete executions? Are they discrete in the sense that a single job which retries should end up under multiple `Execution` rows in the database (so 1 `good_jobs` row, multiple `good_job_executions` rows) and the state is recorded per execution? Or is one `good_job` supposed to have one `good_job_executions` row corresponding to it at all times, and the executions row just gets changed for as long as there are retries? And how should the two libraries _theoretically_ work together, given that:

* A job which is interrupted is not actually erroring - it just (well, "just" is the worst word of them all - "somehow") gets paused to resume afterward. So you don't need to register exceptions for it or set infinite retries (what we had to do for now)
* A job which is interrupted is actually still "the same" job, just with a different `cursor_position` value. Or isn't it? Another way to fix the issue for us was to force a fresh `active_job_id` to be forcibly generated when we re-enqueue, but this does not allow us to see the grouping of the iterations (and the progress of the `cursor_position` value) in the good_job dashboard

Just in case, what we ended up doing is patching `JobIteration::Iteration` like so:

```ruby
module JobIteration::Iteration
  def reenqueue_iteration_job
    ActiveSupport::Notifications.instrument("interrupted.iteration", iteration_instrumentation_tags)
    logger.info("[JobIteration::Iteration] Interrupting and re-enqueueing the job cursor_position=#{cursor_position}")

    adjust_total_time
    self.times_interrupted += 1

    self.already_in_queue = true if respond_to?(:already_in_queue=)
    raise ApplicationJob::RegularInterruption
  end
end
```

and in our `ApplicationJob` resp

```ruby
  # This exception is used in our patch for job-iteration to use exceptions instead of re-enqueues with good_job
  class RegularInterruption < StandardError
  end
  retry_on RegularInterruption, wait: 0, attempts: Float::INFINITY
```

Basically: how is this supposed to work under ideal circumstances? we couldn't really find the right model for identity between the jobs and the executions table, which could inform a fix in either good_job, job-iteration or both libraries. Please advise.

cc @bdewater @Mangara

**Comment by julik** [comment]
Thanks, this explanation makes sense. And what is the difference between `discrete?` vs non-`discrete?` Execution?

WRT what happens currently - well, job-iteration retries the job "as is" after changing some properties of the job, the method looks like this:

```ruby
    def reenqueue_iteration_job
      ActiveSupport::Notifications.instrument("interrupted.iteration", iteration_instrumentation_tags)
      logger.info("[JobIteration::Iteration] Interrupting and re-enqueueing the job cursor_position=#{cursor_position}")

      adjust_total_time
      self.times_interrupted += 1

      self.already_in_queue = true if respond_to?(:already_in_queue=)
      retry_job(wait: @job_iteration_retry_backoff)
    end
```

It does not raise an exception, it gets called inside of `perform` which job-iteration overrides. What we have observed is:

* We enter a check for whether the current execution is an execution for this job
* The same execution (!) gets reset with `finished_at` to nil etc. and saved (!)
* Then outside of that block the `finished_at` gets set again.

It looks like despite the fact that what's happening is clearly a `reenqueued` of `true` (execution.rb line 424) it does not get detected as such. But then it's also not clear from the logic what the objective was:

* When a job is re-enqueued, is the new execution created immediately, to start at some later time?
* Or is the current execution getting re-used?

**Comment by bensheldon** [comment]
I just pushed up #973 which I think will fix this.

> what is the difference between discrete? vs non-discrete? Execution?

Conceptually, a "discrete" execution is an execution record that is tracked separately and distinctly from a job record (i.e. in different database tables). The implementation itself is complicated because I wanted to roll out the database migrations optionally without requiring a major version bump, so we're in a transitional period right now that will get resolved in the next major release.

### Old behavior

- `GoodJob::Job` is a view into the `good_jobs` table that's only used for the Dashboard, and is one-to-one with an Active Job Job.
- `GoodJob::Execution` is a many-to-one relationship to an Active Job Job, and the head/latest record is executed against.

### Current / transitional behavior

- `GoodJob::Job` is (still) a view into the `good_jobs` table that's only used for the Dashboard, and is one-to-one with an Active Job Job.
- `GoodJob::Execution`, when it has the `discrete` boolean set on it, is a one-to-one relationship with an Active Job Job, and it is executed against.
- `GoodJob::DiscreteExecution` is a record of each execution of an Active Job Job

### Future (settled) behavior

- `GoodJob::Job` has a one-to-one relationship with an Active Job Job and it is executed against.
- `GoodJob::Execution` is a record of each execution of the Active Job Job

**Comment by julik** [comment]
I tried the branch you linked - it looks like it helps against the behavior we were observing

**Comment by bensheldon** [comment]
Released in https://github.com/bensheldon/good_job/releases/tag/v3.15.11
